File: python/equity/01-options-pricing/tests/test_black76.py
"""Black-76: equivalence with BSM on the forward, Greeks vs finite differences."""

import math

import pytest

from eq_options import (
    b76_d1_d2,
    black76_greeks,
    black76_price,
    bs_price,
    forward_price,
)

CASES = [
    # (S, K, T, r, sigma, q)
    (100.0, 100.0, 1.0, 0.05, 0.20, 0.00),
    (100.0, 120.0, 0.5, 0.03, 0.35, 0.02),
    (80.0, 60.0, 2.0, -0.01, 0.15, 0.04),
    (250.0, 240.0, 0.1, 0.02, 0.25, 0.01),
]


@pytest.mark.parametrize(("S", "K", "T", "r", "sigma", "q"), CASES)
@pytest.mark.parametrize("otype", ["call", "put"])
def test_black76_equals_bsm_on_model_forward(
    S: float, K: float, T: float, r: float, sigma: float, q: float, otype: str
) -> None:
    """Black-76 with F = S e^{(r-q)T} reproduces BSM to 1e-10."""
    F = forward_price(S, T, r, q)
    assert black76_price(F, K, T, r, sigma, otype) == pytest.approx(
        bs_price(S, K, T, r, sigma, q, otype), abs=1e-10
    )


def test_black76_equals_bs_q_eq_r_when_forward_is_spot() -> None:
    """With q = r the forward equals spot and Black-76(F=S) == BS."""
    S, K, T, r, sigma = 100.0, 105.0, 1.0, 0.04, 0.3
    assert black76_price(S, K, T, r, sigma, "call") == pytest.approx(
        bs_price(S, K, T, r, sigma, r, "call"), abs=1e-12
    )


def test_black76_put_call_parity_on_forward() -> None:
    """C - P = e^{-rT} (F - K)."""
    F, K, T, r, sigma = 2000.0, 1950.0, 0.5, 0.03, 0.18
    c = black76_price(F, K, T, r, sigma, "call")
    p = black76_price(F, K, T, r, sigma, "put")
    assert c - p == pytest.approx(math.exp(-r * T) * (F - K), abs=1e-10)


def test_black76_greeks_vs_finite_differences() -> None:
    F, K, T, r, sigma = 100.0, 95.0, 0.75, 0.04, 0.28
    for otype in ("call", "put"):
        g = black76_greeks(F, K, T, r, sigma, otype)
        h = 1e-5 * F

        def price(f: float = F, t: float = T, rr: float = r, s: float = sigma) -> float:
            return black76_price(f, K, t, rr, s, otype)

        assert g.delta == pytest.approx(
            (price(f=F + h) - price(f=F - h)) / (2 * h), rel=1e-6
        )
        assert g.gamma == pytest.approx(
            (price(f=F + h) - 2 * price() + price(f=F - h)) / h**2, rel=1e-4
        )
        hv = 1e-5
        assert g.vega == pytest.approx(
            (price(s=sigma + hv) - price(s=sigma - hv)) / (2 * hv), rel=1e-6
        )
        ht = 1e-6
        assert g.theta == pytest.approx(
            -(price(t=T + ht) - price(t=T - ht)) / (2 * ht), rel=1e-4
        )
        hr = 1e-6
        assert g.rho == pytest.approx(
            (price(rr=r + hr) - price(rr=r - hr)) / (2 * hr), rel=1e-4
        )


def test_black76_rho_is_minus_T_times_price() -> None:
    g = black76_greeks(500.0, 520.0, 1.5, 0.05, 0.22, "put")
    assert g.rho == pytest.approx(-1.5 * g.price, rel=1e-12)


def test_b76_d1_d2_atm_forward() -> None:
    d1, d2 = b76_d1_d2(100.0, 100.0, 4.0, 0.3)
    assert d1 == pytest.approx(0.3, abs=1e-14)  # sigma*sqrt(T)/2 = 0.3*2/2
    assert d2 == pytest.approx(-0.3, abs=1e-14)


def test_black76_edge_cases() -> None:
    # expiry: intrinsic
    assert black76_price(105.0, 100.0, 0.0, 0.05, 0.2, "call") == 5.0
    # zero vol: discounted intrinsic
    assert black76_price(105.0, 100.0, 1.0, 0.05, 0.0, "call") == pytest.approx(
        5.0 * math.exp(-0.05), abs=1e-12
    )
    with pytest.raises(ValueError):
        black76_price(-1.0, 100.0, 1.0, 0.05, 0.2, "call")

File: cpp/equity-options-engine/tests/test_black76.cpp
// Black-76: exact equivalence with Black-Scholes at F = S e^{(r-q)T},
// forward-space put-call parity, Greeks sanity vs finite differences, and
// edge cases.

#include <gtest/gtest.h>

#include <cmath>
#include <stdexcept>
#include <vector>

#include "eqopt/black76.hpp"
#include "eqopt/black_scholes.hpp"

namespace {

using eqopt::black76_greeks;
using eqopt::black76_price;
using eqopt::bs_price;
using eqopt::forward_price;
using eqopt::OptionType;

TEST(Black76, EqualsBlackScholesOnForwardTo1e12) {
    const std::vector<double> spots{50.0, 100.0, 250.0};
    const std::vector<double> strikes{80.0, 100.0, 130.0};
    const std::vector<double> expiries{0.1, 1.0, 3.0};
    const std::vector<double> vols{0.1, 0.3};
    const std::vector<double> rates{-0.01, 0.04};
    const std::vector<double> divs{0.0, 0.03};
    for (double S : spots)
        for (double K : strikes)
            for (double T : expiries)
                for (double sigma : vols)
                    for (double r : rates)
                        for (double q : divs)
                            for (OptionType type :
                                 {OptionType::Call, OptionType::Put}) {
                                const double F = forward_price(S, T, r, q);
                                const double b76 =
                                    black76_price(F, K, T, r, sigma, type);
                                const double bs =
                                    bs_price(S, K, T, r, sigma, q, type);
                                ASSERT_NEAR(b76, bs,
                                            1e-12 * std::max(1.0, bs))
                                    << "S=" << S << " K=" << K << " T=" << T
                                    << " sigma=" << sigma << " r=" << r
                                    << " q=" << q;
                            }
}

TEST(Black76, ForwardParity) {
    // C - P = e^{-rT} (F - K) to 1e-12.
    const double F = 105.0, K = 98.0, T = 1.4, r = 0.035, sigma = 0.27;
    const double c = black76_price(F, K, T, r, sigma, OptionType::Call);
    const double p = black76_price(F, K, T, r, sigma, OptionType::Put);
    EXPECT_NEAR(c - p, std::exp(-r * T) * (F - K), 1e-12);
}

TEST(Black76, GreeksMatchCentralDifferences) {
    const double F = 100.0, K = 95.0, T = 0.8, r = 0.04, sigma = 0.3;
    for (OptionType type : {OptionType::Call, OptionType::Put}) {
        const auto g = black76_greeks(F, K, T, r, sigma, type);
        const double h_f = 1e-5 * F;
        const double h_v = 1e-5;
        const double h_t = 1e-5 * T;
        const double h_r = 1e-5;
        const auto price = [&](double f, double t, double rr, double sig) {
            return black76_price(f, K, t, rr, sig, type);
        };
        EXPECT_NEAR(g.delta,
                    (price(F + h_f, T, r, sigma) - price(F - h_f, T, r, sigma)) /
                        (2 * h_f),
                    1e-7);
        const double h_f2 = 2e-4 * F;
        EXPECT_NEAR(g.gamma,
                    (price(F + h_f2, T, r, sigma) - 2 * g.price +
                     price(F - h_f2, T, r, sigma)) /
                        (h_f2 * h_f2),
                    1e-7);
        EXPECT_NEAR(g.vega,
                    (price(F, T, r, sigma + h_v) - price(F, T, r, sigma - h_v)) /
                        (2 * h_v),
                    1e-5);
        EXPECT_NEAR(g.theta,
                    -(price(F, T + h_t, r, sigma) - price(F, T - h_t, r, sigma)) /
                        (2 * h_t),
                    1e-5);
        EXPECT_NEAR(g.rho,
                    (price(F, T, r + h_r, sigma) - price(F, T, r - h_r, sigma)) /
                        (2 * h_r),
                    1e-5);
        // rho is pure discounting: -T * V by construction.
        EXPECT_NEAR(g.rho, -T * g.price, 1e-12);
    }
}

TEST(Black76, EdgeCasesAndValidation) {
    // T == 0 -> intrinsic (undiscounted).
    EXPECT_DOUBLE_EQ(black76_price(105.0, 100.0, 0.0, 0.05, 0.2), 5.0);
    // sigma == 0 -> discounted intrinsic.
    EXPECT_DOUBLE_EQ(black76_price(105.0, 100.0, 1.0, 0.05, 0.0),
                     std::exp(-0.05) * 5.0);
    // Daily-margined futures options: r = 0 (no premium discounting).
    const double undisc = black76_price(105.0, 100.0, 1.0, 0.0, 0.2);
    EXPECT_GT(undisc, black76_price(105.0, 100.0, 1.0, 0.05, 0.2));
    EXPECT_THROW(black76_price(-1.0, 100.0, 1.0, 0.05, 0.2),
                 std::invalid_argument);
    EXPECT_THROW(black76_greeks(100.0, 100.0, 0.0, 0.05, 0.2),
                 std::invalid_argument);
}

}  // namespace
